astropy/io/ascii/tests/test_read.py

```diff
@@ -1183,3 +1183,26 @@ def test_no_units_for_char_columns():
     ascii.write(t1, out, format="ipac")
     t2 = ascii.read(out.getvalue(), format="ipac", guess=False)
     assert t2["B"].unit is None
+
+
+def test_initial_column_fill_values():
+    """Regression test for #5336, #5338."""
+
+    class TestHeader(ascii.BasicHeader):
+        def _set_cols_from_names(self):
+            self.cols = [ascii.Column(name=x) for x in self.names]
+            # Set some initial fill values
+            for i in self.cols:
+                i.fill_values = {'--': 0}
+
+    class Tester(ascii.Basic):
+        header_class = TestHeader
+
+    reader = ascii.get_reader(Reader=Tester)
+
+    assert reader.read("""# Column definition is the first uncommented line
+# Default delimiter is the space character.
+apples oranges pears
```

Comment:
That was just copied (and slightly modified) from some docstring. But I've changed it. :+1: